DOC: transform.resize: document output dtype

mdhaber opened the pull request:
## Description

Closes gh-7524

gh-7524 noted that the output dtype of `transform.resize` is not the same as the input dtype. This appears to be an intended feature, so perhaps the issue can be closed by documenting the output dtype. (Feel free to close if there is standard dtype behavior documented elsewhere.)

---

I traced the code a bit, but ultimately determined the relationship between input and output dtype using this script. I varied some of the other parameters manually rather than looping over all possibilities. If you know of another exception, I'd be happy to add it, but I think the consequences of missing an obscure one are rather low. 

```python3
import numpy as np
from skimage.transform import resize

for order in [0, 3, 5]:
    for size in [(41, 41), (50, 50), (57, 57)]:
        for dtype in [np.uint8, np.uint16, np.uint32, np.uint64, 
                    np.int8, np.int16, np.int32, np.int64, 
                    np.float16, np.float32, np.float64]:
            image = np.ones((50, 50), dtype=dtype)
            out = resize(image, size, order=order)
            if out.dtype != np.float64:
                print(order, size, dtype, out.dtype)
```

matthew-brett reviewed (commented):
Looks good.  Suggestion for phrasing.

mkcor commented on skimage/transform/_warps.py:
```diff
@@ -100,6 +100,10 @@ def resize(
         Resized version of the input.
+        The dtype is ``float64`` with the following exceptions:
+
+        - ``order==0``: resized dtype is ``image.dtype``.
```
```suggestion
        - When using `order=0`, output dtype is `image.dtype`.
```

mdhaber replied:
I actually had something like that originally. I changed it for concision.

The backtick "corrections" are incorrect. See https://numpydoc.readthedocs.io/en/latest/format.html#common-rest-concepts.

mkcor replied:
> The backticks are correct as they are. See https://numpydoc.readthedocs.io/en/latest/format.html#common-rest-concepts.

Thanks for sharing this page, @mdhaber. My suggestion is based on: "When referring to a parameter anywhere within the docstring, enclose its name in single backticks" (from https://numpydoc.readthedocs.io/en/latest/format.html#parameters). We've been following this style guide lately, when updating docstrings.

Here, `order` and `image` are parameters, unlike ``float64`` which I agree could/should render in ``monospaced`` font and, hence, be enclosed within ``double backticks``.

mkcor replied:
> I actually had something like that originally. I changed it for concision. Is there risk of confusion as written?

@mdhaber I agree that concision is welcome under a 'Returns' section; my suggestion falls within a broader one about creating a 'Notes' section where, on the contrary, it's fine/good to be verbose.

mdhaber replied:
> My suggestion is based on: "When referring to a parameter anywhere within the docstring, enclose its name in single backticks"

I see how that can still be misunderstood, but it is referring to *just* the name of the parameter, when it appears alone.
I am sure because I wrote the style guide's recommendations regarding backticks (numpy/numpydoc#525).

Through that and /pydata/pydata-sphinx-theme#1852, I've been trying to move the ecosystem's docs to consistently render code in monospaced font so it is easy to distinguish from English.  I proposed to change the recommendation to use double-backticks across the board, even for individual parameter names, but the intent is that someday all these individual parameter names will render as monospaced-font links to the parameter section of the docstring. Feel free to add a thumbs up to numpy/numpydoc#484 if you'd like to see that happen!

---

As for the rest, I can move it if you and Matthew agree.

matthew-brett replied:
Moving is fine with me.

mkcor replied:
> I am sure because I wrote the style guide's new recommendations regarding backticks ([numpy/numpydoc#525](https://github.com/numpy/numpydoc/pull/525)).

Oh, wonderful!! I'm glad I can ask you questions directly then. :grin: (Nice to meet you, @mdhaber!)

> I see how that can still be misunderstood, but it intends to refer to *just* the name of the parameter, when it appears alone.

Aaaah!

> Do you think it needs to be clarified? 

I would say yes, since I was never sure. I would suggest appending "When referring to a parameter anywhere within the docstring, enclose its name in single backticks **if and only if it appears alone**."

> Through that and https://github.com/pydata/pydata-sphinx-theme/issues/1852, I've been trying to move the ecosystem's docs to consistently render code in monospaced font so it is easy to distinguish from English.

I admit that I never 'adhered' to the italic-like rendering of parameter names, since I already use/need this rendering to express emphasis in plain English. I once left a comment about this, somewhere in a PR in this repo. But I wanted to be obedient and consistent...

> I proposed to change the recommendation to use double-backticks across the board, even for individual parameter names,

I would support this. Ouch, we started replacing double backticks which enclosed parameter names with single backticks (not systematically, just when running into them). Now there is a mix of conventions in our docstrings... :grimacing: 

> but the intent is that someday all these individual parameter names will render as monospaced-font links to the parameter section of the docstring. Feel free to add a thumbs up to https://github.com/numpy/numpydoc/pull/484 if you'd like to see that happen!

Nice! Done; incremented the :rocket: emoji.

mkcor commented on skimage/transform/_warps.py:
```diff
@@ -100,6 +100,10 @@ def resize(
+        The dtype is ``float64`` with the following exceptions:
+
+        - ``order==0``: resized dtype is ``image.dtype``.
+        - ``image.dtype`` is ``float16`` or ``float32``: resized dtype is``float32``.
```
```suggestion
        - When `image.dtype` is ``float16`` or ``float32``, output dtype is``float32``.
```

mkcor commented on skimage/transform/_warps.py:
```diff
@@ -100,6 +100,10 @@ def resize(
         Resized version of the input.
+        The dtype is ``float64`` with the following exceptions:
+
+        - ``order==0``: resized dtype is ``image.dtype``.
```
I concur but I would suggest including all this under a 'Notes' section (and refer to it in the 'Returns' section).

mdhaber commented:
Moved this to the Notes and referred to the Notes in the Returns section. Added explanation of why output dtype is floating point and changed exceptions to full sentences. Mentioned `ndimage.zoom` as an alternative that preserves dtype and added to a See Also section.

mkcor commented on skimage/transform/_warps.py:
```diff
@@ -142,6 +146,18 @@ def resize(
     would be [0, 1, 2, 1, 0, 1, 2].
 
+    `resize` uses interpolation. Unless the interpolation method is nearest-neighbor
+    (``order==0``), the algorithm will generate output values as a weighted average
```
```suggestion
    (``order==0``), the algorithm will generate output values as weighted averages
```

mdhaber replied:
I intended for this to be understood as "following a weighted average procedure", so "weighted average" did not refer to an individual numerical result. But sure, this is less likely to raise eyebrows.

> Unrelated to this PR but

This was intended to be a very targeted PR to close gh-7524. I'd rather not expand the scope, if that's alright.

mkcor reviewed (approved):
Thanks for this nice improvement, @mdhaber!

Unrelated to this PR but, after reading the entire docstring, I wonder what "n-dimensional" really means in the context of channels, and whether that's the same as "N-dimensional" (mentioned earlier).

matthew-brett reviewed (approved):
Looks good to me.

mkcor commented:
Wait, @mdhaber did you intend to close this PR? It got two approvals so it was meant to get merged into `main`...

lagru commented:
Reopening – at least for now – so we don't eventually forget this due to its "closed" status.

mdhaber commented:
Apologies; I'm not familiar with scikit-image review policies. If more than two maintainer reviews and a month review period is required, please feel free to leave it open for up to another month. After that, please feel free to use the commits in a separate PR; I'd just rather not have this open for so long. Thanks!
